jigglypuffRL/common/trainer.py

```diff
@@ -0,0 +1,203 @@
+
+    @property
+    def n_envs(self):
+        return self.env.n_envs
+
+
+class OffPolicyTrainer(Trainer):
+    def __init__(self, agent, env, logger, buffer=None, off_policy=True, save_interval=0,
+                 render=False, max_ep_len=1000, distributed=False, ckpt_log_name='experiment',
+                 steps_per_epoch=4000, epochs=10, device='cpu', log_interval=10, batch_size=50,
+                 warmup_steps=1000, start_update=1000, update_interval=50, seed=0):
+        super(OffPolicyTrainer, self).__init__(agent, env, logger, buffer, 
+                                               off_policy, save_interval, render, max_ep_len,
+                                               distributed, ckpt_log_name,steps_per_epoch, 
+                                               epochs, device, log_interval, batch_size, seed
+                                               )
+        self.warmup_steps = warmup_steps
+        self.update_interval = update_interval
+        self.start_update = start_update
+        
+    def train(self):
+        state, episode_reward, episode_len, episode = self.env.reset(), 0, 0, 0
+        total_steps = self.steps_per_epoch * self.epochs
+
+        if 'noise' in self.agent.__dict__ and self.agent.noise is not None:
+            self.agent.noise.reset()
+
+        for t in range(total_steps):
+
+            if t < self.warmup_steps:
+                action = self.env.action_space.sample()
+            else:
+                action = self.agent.select_action(state)
+
+            next_state, reward, done, info = self.env.step(action)
+            if self.render:
+                self.env.render()
+
+            episode_reward += reward
+            episode_len += 1
+            
+            done = False if episode_len == self.max_ep_len else done
+
+            self.buffer.push((state, action, reward, next_state, done))
+
+            states = next_state
+
+            if done or (episode_len == self.max_ep_len):
+                if 'noise' in self.agent.__dict__ and self.agent.noise is not None:
+                    self.agent.noise.reset()
+
+                if episode % self.log_interval == 0:
+                    logger.write({'timestep':t,'Episode':episode,
+                                  'Episode Reward':episode_reward})
+
+                state, episode_reward, episode_len = self.env.reset(), 0, 0
+                episode += 1
+
+            # update params
+            if t >= self.start_update and t % self.update_interval == 0:
+                for _ in range(self.update_interval):
+                    batch = self.buffer.sample(self.batch_size)
+                    states, actions, next_states, rewards, dones = (
+                        x.to(self.device) for x in batch
+                    )
+                    if self.agent.__class__.__name__ == "TD3":
+                        self.agent.update_params(
+                            states, actions, next_states, rewards, dones, t
```

Comment:
Also we should pass `_` instead of `t` here right?

Reply:
Why is the `for _ in range(self.update_interval)` loop there in the first place? I think it should be `for _ in range(self.updates_per_step)`. So the argument to `update_params` should be `t` only right?

Reply:
The difference really is only created with TD3. If we pass `t`, then its hovering between taking updated for 50 steps, and not taking updates for another 50 steps. Whereas if we pass `_`, it takes updates at every alternate step, which is correct.


Note: I've assumed policy update frequency to be 2 here.

Reply:
Okay. Then `_` it is